### PR TITLE
dockerTools.streamLayeredImage: add fakeRootCommands option

### PR DESCRIPTION
**Pull request by vroad**
###### Motivation for this change
Allow adding files/directories owned by users other than root.
You can now optionally specify owner of files by running chown, For example:

```nix
let
  sources = import ./nix/sources.nix;
  overlay = _: pkgs: {
    dockerTools = (import sources.nixpkgs-vroad {}).dockerTools;
  };
  pkgs = import sources.nixpkgs {
    overlays = [ overlay ];
  };
in
  pkgs.dockerTools.buildLayeredImage {
    name = "my-app";
    contents = with pkgs; [
      busybox
    ];
    fakeRootCommands = ''
      mkdir -p ./home/jenkins
      chown 1000:1000 ./home/jenkins
    '';
  }
```

I removed user name to user id mapping of the archive by adding `--numeric-owner` option to tar command, since it didn't seem to be needed for creating a docker image.

https://github.com/NixOS/nixpkgs/issues/94636 talks about implementing similar fakeRoot feature, though OP was talking about adding the feature to buildImage, not buildLayeredImage or streamLayeredImage.

It's even possible to allow running programs other than fakeRoot (by adding more options to `streamLayeredImage` ), though to minimize the scope of the pull request, I didn't include it in the changes.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions (WSL Ubuntu, nixos/nix container image)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -764,18 +768,24 @@ rec {
...
           mkdir $out
 
-          tar \
-            --sort name \
-            --owner 0 --group 0 --mtime "@$SOURCE_DATE_EPOCH" \
-            --hard-dereference \
-            -C old_out \
-            -cf $out/layer.tar .
+          fakeroot bash -c '
+            set -e
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:780]:
```suggestion
          export SHELLOPTS
          fakeroot bash -c '
```

This should make it inherit all shell options.

Comment (roberth) [reply pkgs/build-support/docker/default.nix:780]: I wonder if it's feasible to `source $stdenv/setup` instead.

File: pkgs/build-support/docker/default.nix [hunk 2]
```diff
@@ -738,6 +739,9 @@ rec {
     created ? "1970-01-01T00:00:01Z",
     # Optional bash script to run on the files prior to fixturizing the layer.
     extraCommands ? "",
+    # Optional bash script to run inside fakeroot environment.
+    # Could be used for changing ownership of files in customisation layer.
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:743]: Please (also) document this in the docs: https://github.com/NixOS/nixpkgs/blob/master/doc/builders/images/dockertools.section.md

Same for relevant info from the PR description and discourse message.

**Comment by roberth** [comment]
Interestingly, this [*may*](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=766649) still work on macOS, which apparently has `fakeroot` support too, so it may still [work on darwin](https://github.com/NixOS/nixpkgs/pull/115083) (incl [why](https://github.com/NixOS/nixpkgs/pull/115083#issuecomment-790668250)). Ping @osener

File: pkgs/build-support/docker/default.nix [hunk 3]
```diff
@@ -764,18 +768,24 @@ rec {
         name = "${baseName}-customisation-layer";
         paths = contentsList;
         inherit extraCommands;
+        inherit fakeRootCommands;
```
Comment (SuperSandro2000) [inline pkgs/build-support/docker/default.nix:771]:
```suggestion
        inherit extraCommands fakeRootCommands;
```

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (vroad) [reply pkgs/build-support/docker/default.nix]: That seems better as it makes many utility functions available to fakeRootCommands. I changed that line to `source $stdenv/setup`.

File: pkgs/build-support/docker/default.nix (hunk 2 above)
Comment (vroad) [reply pkgs/build-support/docker/default.nix:743]: Added documentation to that file, though not sure if it's best explanation as I'm not a native speaker of English. Please feel free to suggest changes or correct grammatical mistakes.

**Comment by roberth** [comment]
Could you add a test case?
The test script is in `nixos/tests/docker-tools.nix` and you can add a new image in `pkgs/build-support/docker/examples.nix`.
After that, your PR should be good for merging.

**Comment by vroad** [comment]
> Could you add a test case?
> The test script is in `nixos/tests/docker-tools.nix` and you can add a new image in `pkgs/build-support/docker/examples.nix`.
> After that, your PR should be good for merging.

I added a test that ensures fakeRootCommands has run and set owner of the file to 1000. What do you think?

**Review by roberth — APPROVED** [review]
This is great. Thank you!

**Pull request merged**